pkg/sql/virtual_schema.go

```diff
@@ -385,108 +378,78 @@ func (e virtualDefEntry) getPlanInfo(
+					// will tell us whether or not to let the current row pass the filter.
+					key := constraint.MakeCompositeKey(indexKeyDatums...)
+					span.Init(key, constraint.IncludeBoundary, key, constraint.IncludeBoundary)
+					var err error
+					if idxConstraint.ContainsSpan(p.EvalContext(), &span) {
+						validateRow(datums...)
+						return pusher.pushRow(datums...)
+					}
+					return err
 				}
-				return err
-			}
 
-			giveUp := false
-
-			// We have a virtual index with a constraint. Run the constrained
-			// populate routine for every span. If for some reason we can't use the
-			// index for a span, we set the giveUp flag to true, which will populate
-			// the whole table, filtering the output.
-			for i := 0; i < idxConstraint.Spans.Count(); i++ {
-				span := idxConstraint.Spans.Get(i)
-				if span.StartKey().Length() > 1 {
-					return nil, errors.AssertionFailedf(
-						"programming error: can't push down composite constraints into vtables")
-				}
-				if !span.HasSingleKey(p.EvalContext()) {
-					// No hope - we can't deal with range scans on virtual indexes.
-					giveUp = true
-					break
-				}
-				constraintDatum := span.StartKey().Value(0)
-				// Subtract 2 from the index id to get the ordinal in def.indexes, since
-				// the index with ID 1 is the "primary" index defined by def.populate.
-				virtualIndex := def.indexes[index.ID-2]
-
-				// For each span, run populate.
-				found, err := virtualIndex.populate(ctx, constraintDatum, p, dbDesc,
-					addRowIfPassesFilter)
-				if err != nil {
-					v.Close(ctx)
-					return nil, err
+				// We have a virtual index with a constraint. Run the constrained
+				// populate routine for every span. If for some reason we can't use the
+				// index for a span, we set the giveUp flag to true, which will populate
+				// the whole table, filtering the output.
+				// N.B. we count down in this loop so that, if we have to give up half
+				// way through, we can easily truncate the spans we already processed
+				// from the end and use them as a filter for the remaining rows of the
+				// table.
+				currentConstraint := idxConstraint.Spans.Count() - 1
+				for ; currentConstraint >= 0; currentConstraint-- {
+					span := idxConstraint.Spans.Get(currentConstraint)
+					if span.StartKey().Length() > 1 {
+						return errors.AssertionFailedf(
+							"programming error: can't push down composite constraints into vtables")
+					}
+					if !span.HasSingleKey(p.EvalContext()) {
+						// No hope - we can't deal with range scans on virtual indexes.
+						break
+					}
+					constraintDatum := span.StartKey().Value(0)
+					// Subtract 2 from the index id to get the ordinal in def.indexes, since
+					// the index with ID 1 is the "primary" index defined by def.populate.
+					virtualIndex := def.indexes[index.ID-2]
+
+					// For each span, run populate.
+					found, err := virtualIndex.populate(ctx, constraintDatum, p, dbDesc,
```

Comment:
IDK why, but on a first read I was confused why we were populating the entire table here... Can you just add a quick comment that this will just attempt to do a constrained lookup with the index?